packages/contracts/contracts/core/plugin/AragonUpgradablePlugin.sol

```diff
@@ -0,0 +1,87 @@
+pragma solidity 0.8.10;
+
+import "@openzeppelin/contracts-upgradeable/proxy/utils/UUPSUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import "@openzeppelin/contracts-upgradeable/utils/ContextUpgradeable.sol";
+
+import "../IDAO.sol";
+import "../permission/PermissionManager.sol";
+
+import "../../utils/AppStorage.sol";
+
+/// @title AragonUpgradablePlugin
+/// @author Aragon Association - Giorgi Lagidze - 2022
+/// @notice An Abtract Aragon Plugin(UUPS-UPGRADABLE) that plugin developers have to inherit from.
+abstract contract AragonUpgradablePlugin is Initializable, AppStorage, ContextUpgradeable, UUPSUpgradeable {
+    bytes32 public constant UPGRADE_PERMISSION_ID = keccak256("UPGRADE_PERMISSION");
+
+    /// @notice Emitted only when the plugin gets updated && `update` initialization was called.
+    event PluginUpdated(uint8[3] oldVersion, uint8[3] newVersion, bytes data);
+
+    /// @dev Auth modifier used in all components of a DAO to check the permissions.
+    /// @param _permissionId The hash of the permission identifier
+    modifier auth(bytes32 _permissionId)  {
+        IDAO dao = dao(); 
+        if(!dao.hasPermission(address(this), _msgSender(), _permissionId, _msgData())) {
+            revert PermissionManager.Unauthorized({
+                here: address(this), 
+                where: address(this), 
+                who: _msgSender(), 
+                permissionId: _permissionId
+            });
+        }
+        _;
+    }
+
+    /// @dev Used to check the permissions within the upgradability pattern implementation of OZ
+    function _authorizeUpgrade(address) internal virtual override auth(UPGRADE_PERMISSION_ID) { }
+
+    /// @notice the function logic contract has to include for the custom update logic.
+    /// @dev Dev has a full responsibility whether or not to override it.
+    /// @param oldVersion The old version of logic contract from which contract is upgrading.
+    /// @param newVersion The new version of logic contract to which contract is upgrading.
+    /// @param data the init data for the update
+    function _update(
+        uint8[3] calldata oldVersion, 
+        uint8[3] calldata newVersion, 
+        bytes memory data
+    ) internal virtual;
+    
+    /// @notice allows to call `update` through `reinitializer` modifier if it was not called with this version yet.
+    /// @dev dev can override this and increment it if he/she wants update to be allowed to be called.
+    /// Used also by UI to detect when it can call `update` and when not.
+    /// ex: If the updateVersion(in the new base) > updateVersion(current one), update must be called.
+    /// @return the update version
+    function updateVersion() public virtual returns (uint8) { return 1; }
```

Comment:
This `updateVersion`  doesn't relate exactly to when the updates happen. 

It's possible that contract code has changed, but `updateVersion` still stays whatever number was before. `updateVersion` only gets increased when there's a need to call `update initialization` function and It's super highly unlikely that this would happen more than 256 times..